Add partial docstring support

bennorth opened the pull request:
Assign `__doc__` attribute to functions, methods, classmethods, classes, and modules.

\[Edited: the first version of this PR did not support module docstrings.\]

s-cork commented:
a couple of observations that might help:

```python
class A:
    """docstring"""
    x = __doc__

A.x # 'docstring'

```
Which means that we should really compile this as 
`$loc.__doc__ = $scope123.$const456`
as the first line within the class scope rather than adding it retrospectively after the `buildClass` call.

and for functions the current compile code in this pr for
```python
def foo():
   """bar"""
```

```javascript
$scope156.co_name = new Sk.builtins['str']('foo');
$scope156.co_varnames = [];
$scope156.co_fastcall = 1;
var $funcobj159 = new Sk.builtins['function']($scope156, $gbl);
$funcobj159.$d["__doc__"] = $scope155.$const160;
$loc.foo = $funcobj159;
```

In cpython the doc string of a function is always the first constant in `co_consts`

```python
def foo():
   """bar"""

>>> foo.__code__.co_consts
['bar', None]
```
skulpt doesn't use `co_consts` but I think it would be better to add the docstring to the code obj in someway, before the call to `Sk.builtins.func` which would also make it easier in the future to return the docstring from a `__doc__` descriptor 😄 .

Worth also considering functions that get user defined rather than compiled - so that the `__doc__` would return `None` in this case.

bennorth commented:
Thanks for that feedback, @s-cork.  All good points — I'll look into them and update this PR when done.

bennorth commented:
I've just force-pushed a fresh set of commits, which I think addresses the points from @s-cork:

#### Immediate assignment to `__doc__` in class and module body

This is now done within `Compiler.cbody()`, and the case
```python
class A:
    """docstring"""
    x = __doc__

A.x # 'docstring'
```
now behaves as expected.  It also gives us module docstrings.  The unit tests now cover these cases.

#### Store docstring in code object

> In cpython the doc string of a function is always the first constant in `co_consts` [...] skulpt doesn't use co_consts but I think it would be better to add the docstring to the code obj in someway

I've assigned to `co_docstring` in the code object.  This is currently unused, but as you say it could be used at some point by a descriptor to extract the docstring.

#### User-defined functions

> Worth also considering functions that get user defined rather than compiled

I'm not quite sure what you mean here, sorry.  Do you mean things like `f = lambda x: x+1` (the proposed code gives lambdas a `None` docstring)?  Or closures like
```python
def make_adder(n):
    "Function adding N"
    def add(x):
        "Compute N + X"
        return n + x
    return add
```
(this is from the unit tests)?

s-cork commented:
Sorry I meant that when I do something like...

```javascript
var foo = new Sk.builtin.func(function foo() {}); // i.e. not through the compiler but somewhere in my javascript
```
I also would want it to have a `__doc__` that returns `None`
I could also then do
```javascript
function foo () {};
foo.co_docstring = new Sk.builtin.str('bar')
mod.foo = Sk.builtin.func(function foo() {}); // I now have a __doc__
```

so rather than setting the `$d` in compile code, I was thinking that the func constructor should set the `__doc__` to `co_docstring || None`

If that makes sense.

bennorth commented:
Aha, I see, thanks.  That suggestion simplifies things nicely.  I've pushed an additional commit (plus a test) so you can check whether this does what you mean, and if so, I'll rebase/squash so the PR has a sensible chain of commits.  Let me know what you think.

Would it be worth adding a small helper function to streamline your
```javascript
function foo () {};
foo.co_docstring = new Sk.builtin.str('bar')
mod.foo = Sk.builtin.func(foo); // I now have a __doc__
```
use-case slightly?  Something like
```javascript
Sk.builtin.funcWithDoc = function (func, jsDocstring) {
    func.co_docstring = new Sk.builtin.str(jsDocstring);
    return func;
};
```
so your example could do
```javascript
function foo () {};
mod.foo = Sk.builtin.func(Sk.builtin.funcWithDoc(foo, "bar")); // I now have a __doc__
```
This doesn't gain much readability for a named function, but would avoid having to give an inline function a name just to give it a docstring.  An alternative name would be `withDocstring`.  I'll give it more thought and add a commit providing this, if you like the idea.

s-cork commented on src/compile.js:
```diff
@@ -2145,6 +2152,47 @@ Compiler.prototype.buildcodeobj = function (n, coname, decorator_list, args, cal
+        return "Sk.builtin.none.none$";
+
+    default:
+        Sk.asserts.assert(false, `unexpected node kind ${node.constructor}`);
```
you can use `Sk.asserts.fail(msg)` here

bennorth replied:
Thanks; didn't spot that one!

s-cork commented on src/compile.js:
```diff
@@ -2683,8 +2731,19 @@ Compiler.prototype.exitScope = function () {
+    // happens for class and module bodies.
+    //
+    const maybeDocstring = this.maybeCDocstringOfBody(stmts);
+    if (maybeDocstring != null) {
```
we can use `!==` here

s-cork commented on src/compile.js:
```diff
@@ -2145,6 +2152,47 @@ Compiler.prototype.buildcodeobj = function (n, coname, decorator_list, args, cal
+ * body has no docstring.
+ */
+Compiler.prototype.maybeCDocstringOfBody = function(body) {
+    if (body.length == 0)  // Don't think this can happen?
```
suggestion `===`

s-cork commented on src/compile.js:
```diff
@@ -2145,6 +2152,47 @@ Compiler.prototype.buildcodeobj = function (n, coname, decorator_list, args, cal
+    if (body.length == 0)  // Don't think this can happen?
+        return null;
+
+    const stmt_0 = body[0];
+    if (stmt_0.constructor !== Sk.astnodes.Expr)
+        return null;
+
+    const expr = stmt_0.value;
+    if (expr.constructor !== Sk.astnodes.Str)
+        return null;
```
in the style guide we should add `{` but compile.js is ignored for es linting

bennorth replied:
I've left it without the `{}`s to keep it concise; happy to put them in if preferred.

s-cork reviewed (approved):
lgtm - nice pr - seems very closely aligned to cpython - @skulpt/committers ready for you

one other thing you might want to look at is in `type.js`
```javascript
        // set __module__ if not present
        if(dict.mp$lookup(Sk.builtin.str.$module) === undefined) {
            dict.mp$ass_subscript(Sk.builtin.str.$module, Sk.globals["__name__"]);
        }
```
You could do something similar for `__doc__` so that even a class that doesn't define a `__doc__` still gets `None`

s-cork commented:
re the helper function - I think that's beyond the scope of this pr but an interesting discussion point and presents a wider question about our api for instantiating a `new Sk.builtin.func` with the correct `__name__` and `__doc__` from javascript. See #1014 where I think the api is also mentioned. Currently if you try to print most of the builtin funcs you get the `<Native JS>` rather than the actual name of the object. 

in my head it would be something like this. 

```javascript
mod.foo = new Sk.builtin.func(new Sk.builtin.codeobj(() => {
    /// funciton body
}, {
    name: "foo",
    doc: "bar",
    //... etc
}));
```

bennorth commented:
Thanks for further feedback.  I've made the individual changes you suggest, leaving out the `{}`s on the `return null`s.  I also now assign to `__doc__` of a class if there is no docstring given.  New set of commits force-pushed.

bennorth commented:
Agreed that an API for adding a docstring when calling `Sk.builtin.func()` would be scope-creep here — thanks for writing the idea up (#1263).

bnmnetp commented:
@bennorth it seems the PR I accepted yesterday has caused conflicts in several good PRs.  I would like to get this accepted if you can resolve these...

bennorth commented:
@bnmnetp The machinery added by #1092 made a few of the changes in this PR unnecessary, so I thought it would be cleaner to create a separate PR — #1265 — for the docstring feature, now that #1092 is in.  If you're happy with that approach, I can close this PR.

bnmnetp commented:
Nice!  That does simplify this PR considerably!

You can go ahead and close this one.

BTW, I tried to have a look at Pytch but could not get the tutorial to work.  Something in the console about an unhandled Promise.  This seems like it could be an interesting addition to the RunestoneInteractive project.

bennorth commented:
@bnmnetp Thanks; will do.

(And many thanks for looking at Pytch.  I don't want to hijack this PR for Pytch troubleshooting, but: We know that it doesn't work under Safari, and will investigate as resources allow.  Meanwhile, if you had time for another look using Chrome or Firefox, that would be great!  I hadn't come across Runestone Interactive before now — its goals are excellent, and I will read about it in more detail.)
